### DISCUSSION TITLE
How to lift a pure value into the effect-space inside an async function

### DISCUSSION THREAD

**Discussion by dirkil** [louthy/language-ext#1336] [Q&A] [answered]
Hi there,

I have the following function that creates an offer. In case the call succeeds it returns the `Guid` of the created offer, otherwise an `Error`. Here is the code for that:
```csharp
Aff<Guid> CreateOffer(Guid mandantId, string offerJson)
{
        return Aff(async () =>
        {
            var content = new StringContent(immobilieJson, Encoding.UTF8, "application/json");
            var response = await _httpClient.PostAsync($"api/offers/{mandantId}", content);
            if (response.StatusCode == HttpStatusCode.OK)
            {
                return await response.Content.ReadFromJsonAsync<Guid>();
            }

            var errorContent = await response.Content.ReadAsStringAsync();

            var error = Error.New((int)response.StatusCode, errorContent);
            return FailAff<Guid>(error);
        });
}
```
Unfortunately, this code does not compile, and it has to do with the last return statement and/or the `return Aff( async() =>...)` at the beginning. The compilation error is:

_Error CS0411 : The type arguments for method Prelude.Aff<RT, A>(Func<RT, ValueTask<A>>) cannot be inferred from the usage._

As far as I understand it, I have to lift up a pure value into the effect space which I thought that `FailAff` does for me. I guess it has to do with the fact that the method  must return a `ValueTask` but I have no idea how to do that.

Can anyone help me, please?

**Comment by harrhp** [comment, marked as answer]
`Prelude.Aff` can't return non-exceptional errors, you should use `Prelude.AffMaybe` for that. It expects `Fin<A>` as return value, but `Prelude.FailAff` returns `Aff<A>`, so you have to use `Prelude.FinFail` instead

```
Aff<Guid> CreateOffer(Guid mandantId, string offerJson)
{
  return AffMaybe(async () =>
  {
    var content = new StringContent(immobilieJson, Encoding.UTF8, "application/json");
    var response = await _httpClient.PostAsync($"api/offers/{mandantId}", content);
    if (response.StatusCode == HttpStatusCode.OK)
    {
      return await response.Content.ReadFromJsonAsync<Guid>();
    }
    var errorContent = await response.Content.ReadAsStringAsync();
    var error = Error.New((int)response.StatusCode, errorContent);
    return FinFail<Guid>(error);
  });
}
  ```

**Reply by dirkil** [reply]
Hi harrhp,

thank you very much for your answer. It solved my problem! And many thanks for the explanation. Now I know what `AffMaybe` is for, I have always wondered...
